internal/newtelemetry/client.go

```diff
@@ -0,0 +1,344 @@
+	}
+	return c.distributions.LoadOrStore(namespace, name, tags)
+}
+
+func (c *client) ProductStarted(product Namespace) {
+	c.products.Add(product, true, nil)
+}
+
+func (c *client) ProductStopped(product Namespace) {
+	c.products.Add(product, false, nil)
+}
+
+func (c *client) ProductStartError(product Namespace, err error) {
+	c.products.Add(product, false, err)
+}
+
+func (c *client) RegisterAppConfig(key string, value any, origin Origin) {
+	c.configuration.Add(Configuration{key, value, origin})
+}
+
+func (c *client) RegisterAppConfigs(kvs ...Configuration) {
+	for _, value := range kvs {
+		c.configuration.Add(value)
+	}
+}
+
+func (c *client) Config() ClientConfig {
+	return c.clientConfig
+}
+
+// Flush sends all the data sources before calling flush
+// This function is called by the flushTicker so it should not panic, or it will crash the whole customer application.
+// If a panic occurs, we stop the telemetry and log the error.
+func (c *client) Flush() {
+	defer func() {
+		r := recover()
+		if r == nil {
+			return
+		}
+		log.Warn("panic while flushing telemetry data, stopping telemetry: %v", r)
+		telemetryClientDisabled = true
+		if gc, ok := GlobalClient().(*client); ok && gc == c {
+			SwapClient(nil)
+		}
+	}()
+
+	payloads := make([]transport.Payload, 0, 8)
+	for _, ds := range c.dataSources {
+		if payload := ds.Payload(); payload != nil {
+			payloads = append(payloads, payload)
+		}
+	}
+
+	nbBytes, err := c.flush(payloads)
+	if err != nil {
+		log.Warn("telemetry: error while flushing: %v", err)
+	}
+
+	if c.clientConfig.Debug {
+		log.Debug("telemetry: flushed %d bytes of data", nbBytes)
+	}
+}
+
+func (c *client) transform(payloads []transport.Payload) []transport.Payload {
+	c.flushMapperMu.Lock()
+	defer c.flushMapperMu.Unlock()
+	payloads, c.flushMapper = c.flushMapper.Transform(payloads)
+	return payloads
+}
+
+// flush sends all the data sources to the writer after having sent them through the [transform] function.
+// It returns the amount of bytes sent to the writer.
+func (c *client) flush(payloads []transport.Payload) (int, error) {
+	payloads = c.transform(payloads)
+
+	if c.payloadQueue.IsEmpty() && len(payloads) == 0 {
+		return 0, nil
+	}
+
+	emptyQueue := c.payloadQueue.IsEmpty()
+	// We enqueue the new payloads to preserve the order of the payloads
+	c.payloadQueue.Enqueue(payloads...)
+	payloads = c.payloadQueue.Flush()
+
+	var (
+		nbBytes        int
+		speedIncreased bool
+		failedCalls    []internal.EndpointRequestResult
+	)
+
+	for i, payload := range payloads {
+		results, err := c.writer.Flush(payload)
+		c.computeFlushMetrics(results, err)
+		if err != nil {
+			// We stop flushing when we encounter a fatal error, put the bodies in the queue and return the error
+			if results[len(results)-1].StatusCode == 413 { // If the payload is too large we have no way to divide it, we can only skip it...
+				log.Warn("telemetry: tried sending a payload that was too large, dropping it")
+				continue
+			}
```

Comment:
We _could_ drop any payload for which we receive an HTTP 4XX response (logging that).
It might also be useful to specify what kind of payload is getting dropped so it's easier to track down when that happens.